newton/_src/sim/builder.py

```
@@ -484,6 +486,203 @@ def __init__(self, up_axis: AxisType = Axis.Z, gravity: float = -9.81):
+            existing = self.custom_attributes[full_key]
+            if (
+                existing.frequency != frequency
+                or existing.dtype != dtype
+                or existing.assignment != assignment
+                or existing.namespace != namespace
+            ):
+                raise ValueError(
+                    f"Custom attribute '{full_key}' already exists with frequency='{existing.frequency}', "
+                    f"dtype='{existing.dtype}', assignment='{existing.assignment}', namespace='{existing.namespace}'. "
+                    f"Cannot redefine with frequency='{frequency}', dtype='{dtype}', assignment='{assignment}', namespace='{namespace}'."
+                )
```

Ruff (0.14.1)

Avoid specifying long messages outside the exception class

(TRY003)